### DISCUSSION TITLE
Create a package-lock.json file from pnpm-lock.yaml

### DISCUSSION THREAD

**Discussion by maclanzallamas** [pnpm/pnpm#3367] [Q&A]
It's clear how to do this the other way around with pnpm import (https://pnpm.io/cli/import)
But if I wanted to migrate from pnpm to npm, how do I use the _pnpm-lock.yaml_ to create a _package-lock.json_ file?

Similar question: https://github.com/pnpm/pnpm/discussions/3356

Thanks

**Comment by frank-dspeed** [comment]
I am searching for the same answer i guess the most easy way is to do simply the migration to npm workspaces then do a npm install and drop pnpm the only compatible file is the package.json you can shim your dev environment like the one from pnpm with npm overrides. 

That gives you the needed flexebility i am still exploring the patterns so i have no full workflow for that but it is in progress.

**Reply by bendman** [reply]
The ability to generate the package-lock file would also be useful for wiring other tools that use a package-lock.json, such as dependabot.

**Reply by frank-dspeed** [reply]
@bendman final conclusion is pnpm philosophie is that npm should integrate that if they want that they do not depend on package-lock.json so they see no need for interfaces and tooling even if contributed from outside.

my final conclusion is pnpm is gone wild and i will not use it anymore while i liked some design patterns they did use i do not agree with everything so the best is to simply go away from the monolit pnpm to something more modular managing dependencies with npm got much better with the last releases.

see: https://github.com/pnpm/pnpm/issues/4485

**Comment by zkochan** [comment]
If you need this feature, you need to ask for it in the npm CLI repository. Why would we spend our time on a feature that helps people switch away from pnpm to another package manager? It doesn't make sense.

**Reply by frank-dspeed** [reply]
@zkochan nope pnpm makes no sense if it does vendor lockin pnpm should be a utility that helps and it should never get a blocker but i accept your philosophy it is ok to think like that but the only thing that leads to is Hardforking pnpm into something diffrent like pnpm-utils

and then you will loose all your consumers anyway as they will go to the better more compatible product anyway In the Internet the Winner always takes it all.

**Reply by alexherrera42** [reply]
This feature isn't needed to run away to another package manager.  This feature is needed to be able to decouple individual packages from the monorepo for distribution.   I don't want to include the entire monorepo in each container image.  I also don't want to have to install all the node modules for the entire monorepo in each container image.  Why do you not have an export feature that will generate a self contained package with a lockfile(even a pnpm lockfile) that ONLY includes the packages relevant to that package that can just be loaded straight into a container image?  Why do I find myself having to muck about with rsync filters to do so by hand?  pnpm works great in dev, but trying to build a reasonable pipeline to get my packages distributed into lean individualized containers is a nightmare.

The pack utility for pnpm is useless because it doesn't generate a package specific lock file.   I could tarball the package folder just fine without help from pnpm.

We don't want to run to npm.  We want to use pnpm.  But pnpm isn't offering the full solution here.  We need a way to ship individual packages out without dragging along the whole monorepo and all other package dependencies along for the ride.

**Reply by zkochan** [reply]
You need something like [a deploy command](https://github.com/pnpm/pnpm/issues/4378). Generating a `package-lock.json` is not a solution to your problem.

**Reply by alexherrera42** [reply]
It is A solution.  Skimming that link you posted this does seem like it may offer some promise, and I agree that would be a more robust and complete solution.  Though it looks as though this is not currently under development?

**Reply by zkochan** [reply]
It is the right solution. It doesn't make sense to use one package manager locally and another one in CI or docker.

I suggested a working solution in that issue: https://github.com/pnpm/pnpm/issues/4378#issuecomment-1074586535

**Reply by alexherrera42** [reply]
Indeed, I saw that working solution, and I intend to give it a whirl.  The goal isn't to use 2 package managers though.  As I stated, using npm would be a workaround.  It's desirable to be able to do it all with pnpm.   If @pnpm/make-dedicated-lockfile would work without workspace dependencies being required to exist in the npm registry that would be highly beneficial.  Really there are two issues, one I mentioned above(getting a package fully isolated for containerization), and the other being the necessity for some kind of change signal to allow for multistage building.  It is desirable to be able to tell if node_modules has changed since the last container build.  If it has, then we need to build a new intermediate container filled with node_modules.   If not, we can skip that step and the next stage can just use the :latest for that intermediate container.   One way to do this would be to cache and check the lock file.   The problem is pnpm-lock.yaml will signal change for ANY package's node_modules being changed since it stores state for all packages in the monorepo.  This is not desirable.  We want to be able to tell when a specific package and that specific package only has had its modules changed.  For this we would need to be able to generate a dedicated lock file which only has the state for the node modules of the specific package we are containerizing.

As a shoddy workaround I'm currently playing with just storing the output of pnpm ls --prod to be cached and checked against to signal change.  It seems to be functional so far, but it's pretty hacky.

**Reply by matzeeable** [reply]
Hey @alexherrera42 ! Did you already found a solution for this? I found a promising tool, but I have found another bug: https://github.com/Madvinking/pnpm-isolate-workspace/issues/14

**Reply by alexherrera42** [reply]
The solution I have employed utilizes rsync with filters to copy the monorepo directory into the container with just the target package and pnpm lock file.  Then it runs "pnpm i --prod --filter [package name]" in the container.  My rsync filter file looks like this:
```
- node_modules/
+ pnpm-lock.yaml
+ pnpm-workspace.yaml
+ package.json
+ /packages/
+ somepackagedirectory/
- *
```
if you've never used rsync filters before they can be a little tricky.  Definitely read the docs: https://linux.die.net/man/1/rsync

**Reply by alexherrera42** [reply]
I should also mention that the above described process produces a container with just the node modules for the selected package(s).  This is part of a multistage build, and I have another container build that uses this node_modules container as a base and uses rsync to bring just the dist folder of the built package in.

**Reply by DonKirkham** [reply]
Here is my use-case: while I really like pnpm and use it quite a bit, not having the ability to generate the equivilant package-lock.json files means anyone that wants to use my solutions is also locked into pnpm. I would like to offer them the option to use npm if that is their choice. I currently am working on a project the works fine if I use pnpm, but if I do a install with npm (no lock file available because I used pnpm), I get multiple errors that seem to be related to different versions of some dependancies. Now I have to figure that out to continue using this project (for training). I don't have time for that. I already have to deal with some projects not working unless "--shamefully-hoist" is used and this is just another hurdle to using pnpm.

So, the answer to your question about why: this drives away users. As a speaker at many conference where I present on open source solution building, I find it harder and harder to recommend pnpm. I am probably going to abandon it and stick to npm until I see that the benefits of pnpm outweigh the challenges.

**Reply by GorillaCoder** [reply]
Hi @zkochan, you say:
```Why would we spend our time on a feature that helps people switch away from pnpm to another package manager? It doesn't make sense.```

You wouldn't. You shouldn't. You need to get folks to adopt pnpm! 100% agree. Thing is, not supporting export to package-lock.json hurts pnpm. Slows its adoption. And, actually, doesn't do much to stop folks switching away from pnpm. It just pisses them off making it more likely they will leave.

Here's the thing. Our dev pipeline uses other tools. Projects like snyk. Projects that have limited resources and have deprioritized supporting pnpm. Why would they prioritize it, the pnpm community is small. How do you convince them to support you. By growing your community. And, much as you may not like it, that means supporting export to package-lock.json. You worry that supplying this feature will diminish your user base. I, and others here, are telling you it will ease our migration to pnpm. Want to grow your user base: support export to package-lock.json.

When you have crossed the tipping point, and are big enough that all the tools vendors have to support you, that's when you stop supporting export. Until then, not supporting export hurts pnpm

**Reply by weyert** [reply]
You make the contributions to those projects yourself, if you are eager to use PNPM and those tools. Or alternatively switch to a different tool that does support PNPM already. Or sponsor @zkochan to do this development work

**Reply by jamesmortensen** [reply]
I would love to be able to use PNPM more to save disk space and not keep repeatedly downloading the same node_modules to N different projects.  Unfortunately, it feels clunky to use it without the ability to have it work with the tools my colleagues use.  PNPM is a way better package manager than the other two major options.  Working alone, I'll use PNPM, but I can't use it for any serious collaborative work, which is really a shame.  We just can't get groups of people to all agree, so we adapt.  I adapt by using NPM instead of PNPM, because managing the disk space is easier than dealing with the package-lock.json issue.

Unfortunately, `npm i --package-lock-only` didn't quite work for me on one of our legacy projects where there were some funky things happening with peer dependencies.

I get it. It's difficult for open source maintainers to prioritize everything. So the bigger question is would y'all be willing to accept a pull request that implements what folks are looking for?

**Reply by weyert** [reply]
That's a question for @zkochan to answer. Personally, I don't have issues with it but still think PNPM support needs to be added to the tooling (when open-source).

**Reply by zkochan** [reply]
> Why would they prioritize it, the pnpm community is small.

It is not small.

If they decide not to support pnpm, they hurt themselves more than they hurt this project.

> Or sponsor @zkochan to do this development work

Sponsoring is nice but I still decide what are my priorities. I don't see value in this. If someone wants to work on it, fine. But I don't want to.

**Comment by jamesmortensen** [comment]
It looks like this already may be possible, unless I'm misunderstanding the exact use case:

Generate a package-lock.json file without installing dependencies in node_modules:
`npm i --package-lock-only`

I was only using pnpm for demo projects to save disk space because I needed a package-lock.json on the CI servers, but it looks like I can just generate it via npm while keeping the node_modules symlinked to a central store.  I think this means I can now use pnpm for work projects too.  

See https://stackoverflow.com/a/54910009/552792

**Reply by frank-dspeed** [reply]
Amazing job and finding i need to document that you saved a lot of time your the hero of the day at present !!!

**Reply by PythonCoderAS** [reply]
This should be the accepted answer.

**Reply by frank-dspeed** [reply]
@PythonCoderAS it is already the accepted answer it works like a charm pnpm is now free and even useable with npm together via workspaces

```
mkdir project
cd project
npm init
npm init -w ./workspaces/pnpm-stuff
cd workspaces/pnpm-stuff
pnpm init
pnpm install
npm i --package-lock-only
cd ../..
npm install ## Will install and link all other workspaces and your current project including pnpm-stuff
```

So you can manage pnpm stuff inside pnpm and keep working with regular packages on projects that are not pnpm easy interop.

**Reply by zkochan** [reply]
With this solution you will have completely different packages in `package-lock.json` and `pnpm-lock.yaml`.

**Reply by frank-dspeed** [reply]
@zkochan thats correct after that point the package-lock.json is again the authoring format and pnpm-lock needs to get recreated from the package.json and package-lock.json but as users most of the time will have NPM at the root of the project and only do need pnpm as addition when they build and install some deps that depend on that its fine.

Most people do Only install to build there are less situations where you will ship the node_modules folder in production.

**Comment by rsheasby** [comment]
Sorry to necro this discussion but I wanted to add another POV. This feature would be very useful for my team not because we want to switch away from pnpm or because we want to do some unsupported deploy wizardry. On the contrary, this would be very useful because we **don't want to switch away from pnpm**, but some of our other tools don't support pnpm.

Specifically the issue we're having is with dependency scanning tools like snyk, semgrep, etc. These tools need to read a npm/yarn lock file in order to scan dependencies, but do not support pnpm, thus forcing us to choose between pnpm or dependency scanning (or using a clunky workaround, which is what we're currently investigating)

**Reply by weyert** [reply]
Why not spend the time to contribute to these tooling and add PNPM support? My work is on Gitlab and they don't support PNPM either in some of the compliance tooling. For this reason, I recently contributed the `licenses`-command as that is one of the requirements by compliance to be able to use PNPM. Currently working with `license-finder` to add support for PNPM with this new command. Next stop, is adding support for PNPM in the vulnerabilities scanner  Gemnasium used by Gitlab. If `snyk` supports SPDX format as input then you could consider trying out Gemnasium.

If you are happen to be using Gitlab you will soon have the PNPM support :)

Personally, I don't see the need to add support for generating `package-lock.json` files that energy can be better spend else where in the project.

**Reply by rsheasby** [reply]
> Why not spend the time to contribute to these tooling and add PNPM support?

I am working on exactly this with one of my vendors to add pnpm compatability, but it's significant effort for both teams because we do not know pnpm internals very well, and we each have higher priority items.

> that energy can be better spend else where in the project

Sure, but you must understand that this merely offloads the energy to every user of your product to find workarounds, and every other tool that exists which doesn't have a large enough development team to justify spending effort to support a (relatively) uncommon package manager. Ultimately it's absolutely the choice of this project's maintainers to reject this feature request, but make no mistake: this is at the cost of people who want to use the project, and likely hinders adoption.

Adding the ability to output a package-lock.json would be a single fix that enables compatibility with dozens/hundreds of other projects, and reduces the duplicated effort required to add support to each individual project. It's a compatability feature that IMO is useful for many people.

Like I said, if this is something the pnpm team is unwilling to consider then that's entirely fine, but I honestly believe it comes at the cost of pnpm adoption.

**Reply by weyert** [reply]
Yeah, well, I personally I think it hinders adaption of those services if they don't support PNPM. Typically, they do take the time to support Yarn. 

Does Yarn allow you to generate package-lock.json files?

I can understand for open-source products, but if I pay thousands of dollars per month for things like Snyk or Gitlab. I would expect they should support the major package managers.  Creating a `package-lock.json` feature in PNPM also wouldn't help the increase the support for PNPM in such tooling. Which tooling are you exactly using that doesn't have PNPM lock file reading support? 

I would hope that Snyk at least supports the SPDX / CycloneDX format that quite useful for SBOM etc.

**Reply by weyert** [reply]
Also we are using `semgrep` via Gitlab too but I am not aware it does any specific `yarn` or `npm` checks? What's the current unsupported functionality for `semgrep`? You might be in luck if it is needed to be able for _me_ to use *pnpm* at work.

**Reply by rsheasby** [reply]
Are you using FOSS semgrep or Semgrep Supply Chain? Supply Chain does not support pnpm (which would be the only instance where you need to read the package list)

**Reply by weyert** [reply]
Probably Gitlab uses FOSS semgrep. Gitlab uses this as their vulnerability scanning:
https://gitlab.com/gitlab-org/security-products/analyzers/gemnasium

Working on adding PNPM support

**Comment by weyert** [comment]
`license-finder` just added support for PNPM:
https://github.com/pivotal/LicenseFinder/pull/954

**Comment by weyert** [comment]
Just now I found https://github.com/oss-review-toolkit/ort appears to support PNPM
